File: source/go/kruskal_mst.go
package main

import "fmt"

var ufParent []int
var ufRank []int

func ufInit(n int) {
	ufParent = make([]int, n)
	ufRank = make([]int, n)
	for i := 0; i < n; i++ {
		ufParent[i] = i
	}
}

func ufFind(x int) int {
	for ufParent[x] != x {
		ufParent[x] = ufParent[ufParent[x]]
		x = ufParent[x]
	}
	return x
}

func ufUnion(a, b int) bool {
	ra := ufFind(a)
	rb := ufFind(b)
	if ra == rb {
		return false
	}
	if ufRank[ra] < ufRank[rb] {
		ufParent[ra] = rb
	} else if ufRank[ra] > ufRank[rb] {
		ufParent[rb] = ra
	} else {
		ufParent[rb] = ra
		ufRank[ra]++
	}
	return true
}

type Edge struct {
	u, v, w int
}

func main() {
	n := 5
	edges := []Edge{
		{0, 1, 2},
		{0, 3, 6},
		{1, 2, 3},
		{1, 3, 8},
		{1, 4, 5},
		{2, 4, 7},
		{3, 4, 9},
	}

	// Insertion sort by weight
	for i := 1; i < len(edges); i++ {
		keyEdge := edges[i]
		j := i - 1
		for j >= 0 && edges[j].w > keyEdge.w {
			edges[j+1] = edges[j]
			j--
		}
		edges[j+1] = keyEdge
	}

	ufInit(n)

	var mstEdges []Edge
	total := 0

	for _, e := range edges {
		if ufUnion(e.u, e.v) {
			mstEdges = append(mstEdges, e)
			total += e.w
		}
	}

	fmt.Println("kruskal MST edges:")
	for _, e := range mstEdges {
		fmt.Printf("  %d - %d (weight %d)\n", e.u, e.v, e.w)
	}
	fmt.Printf("total weight: %d\n", total)
}

File: source/python/kruskal_mst.py
uf_parent = []
uf_rank = []

def uf_init(n):
    global uf_parent, uf_rank
    uf_parent = list(range(n))
    uf_rank = [0] * n

def uf_find(x):
    while uf_parent[x] != x:
        uf_parent[x] = uf_parent[uf_parent[x]]
        x = uf_parent[x]
    return x

def uf_union(a, b):
    ra = uf_find(a)
    rb = uf_find(b)
    if ra == rb:
        return False
    if uf_rank[ra] < uf_rank[rb]:
        uf_parent[ra] = rb
    elif uf_rank[ra] > uf_rank[rb]:
        uf_parent[rb] = ra
    else:
        uf_parent[rb] = ra
        uf_rank[ra] += 1
    return True

n = 5
edges = [
    {"u": 0, "v": 1, "w": 2},
    {"u": 0, "v": 3, "w": 6},
    {"u": 1, "v": 2, "w": 3},
    {"u": 1, "v": 3, "w": 8},
    {"u": 1, "v": 4, "w": 5},
    {"u": 2, "v": 4, "w": 7},
    {"u": 3, "v": 4, "w": 9},
]

edges.sort(key=lambda e: e["w"])
uf_init(n)

mst_edges = []
total = 0
for edge in edges:
    if uf_union(edge["u"], edge["v"]):
        mst_edges.append(edge)
        total += edge["w"]

print("kruskal MST edges:")
for e in mst_edges:
    print(f"  {e['u']} - {e['v']} (weight {e['w']})")
print(f"total weight: {total}")
